packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -248,33 +248,40 @@ export class LangiumGrammarValidator {
         }
         for (const rule of grammar.rules.filter(ast.isParserRule)) {
             const isDataType = isDataTypeRule(rule);
-            if (!isDataType && rule.type?.name && types.has(rule.type.name) === !!rule.infers) {
-                const keywordNode = rule.infers ? findKeywordNode(rule.$cstNode, 'infers') : findKeywordNode(rule.$cstNode, 'returns');
-                accept('error', getMessage(rule.type.name, rule.infers), {
-                    node: rule.type,
+            const isInfers = !!rule.inferredType;
+            const ruleTypeName = getTypeName(rule);
+            if (!isDataType && ruleTypeName && types.has(ruleTypeName) && isInfers) {
+                const keywordNode = isInfers ? findKeywordNode(rule.$cstNode, 'infer') : findKeywordNode(rule.$cstNode, 'returns');
```

Comment:
Using the variable `isInfers` makes handling the `returns` case impossible: if `isInfers` is false, the evaluation never enter this if-branch and achieves the `findKeywordNode(rule.$cstNode, 'returns')` part.